common/environment/environment.cpp

```diff
@@ -181,27 +181,43 @@ class CLocalEnvironment : implements IConstEnvironment, public CInterface
 
                 keyPairMap[name] = { absPublicKeyPath.str(), absPrivateKeyPath.str() };
             }
-            Owned<IPropertyTreeIterator> clusterIter = p->getElements("EnvSettings/Keys/Cluster");
-            ForEach(*clusterIter)
+
+/* From 7.0.0 until 7.0.6, the <Keys> section of the environment required
+ * the mappings to be defined as "Cluster" instead of "ClusterGroup" - See: HPCC-21192
+ */
+#define BKWRDCOMPAT_CLUSTER_VS_CLUSTERGROUP
+
+            const char *groupKeysPath = "EnvSettings/Keys/ClusterGroup";
+
+#ifdef BKWRDCOMPAT_CLUSTER_VS_CLUSTERGROUP
+            for (unsigned i=0; i<2; i++) // once for std. "ClusterGroup", 2nd time for legacy "Cluster"
             {
-                IPropertyTree &cluster = clusterIter->query();
-                const char *clusterName = cluster.queryProp("@name");
-                if (isEmptyString(clusterName))
-                {
-                    WARNLOG("skipping EnvSettings/Keys/Cluster entry with no name");
-                    continue;
-                }
-                if (cluster.hasProp("@keyPairName"))
+#endif
+                Owned<IPropertyTreeIterator> clusterIter = p->getElements(groupKeysPath);
+                ForEach(*clusterIter)
                 {
-                    const char *keyPairName = cluster.queryProp("@keyPairName");
-                    if (isEmptyString(keyPairName))
+                    IPropertyTree &cluster = clusterIter->query();
+                    const char *clusterName = cluster.queryProp("@name");
+                    if (isEmptyString(clusterName))
                     {
-                        WARNLOG("skipping invalid EnvSettings/Key/Cluster entry, name=%s", clusterName);
+                        WARNLOG("skipping %s entry with no name", groupKeysPath);
                         continue;
                     }
-                    keyClusterMap[clusterName] = keyPairName;
+                    if (cluster.hasProp("@keyPairName"))
+                    {
+                        const char *keyPairName = cluster.queryProp("@keyPairName");
+                        if (isEmptyString(keyPairName))
+                        {
+                            WARNLOG("skipping invalid %s entry, name=%s", groupKeysPath, clusterName);
+                            continue;
+                        }
+                        keyClusterMap[clusterName] = keyPairName;
+                    }
                 }
+#ifdef BKWRDCOMPAT_CLUSTER_VS_CLUSTERGROUP
+                groupKeysPath = "EnvSettings/Keys/Cluster";
```

Comment:
Should we always do this ?  Or at least until we build with BKWRDCOMPAT_CLUSTER_VS_CLUSTERGROUP undefined ?
I mean if we find keys in ClusterGroup - should we still continue to on and overwrite them with what is found in Cluster ?

Reply:
This backward compat. is to allow either, but in particular to allow for old environments that just have old "Cluster" definitions in, to continue to work.

Anyone who has mixed definitions, neither has a purely legacy definition, nor a new corrected definition, so it should probably just error at that point, either way they are asking for trouble.

There's also nothing to stop them adding multiple definitions that override one another, e.g. later "ClusterGroup"'s will override earlier "ClusterGroup"'s. 

I think it's fine as is and hopefully we can delete it in near future.

Reply:
I would reverse it then - read the old first and then overwrite with what is new.  Seems safer.

Reply:
I agree if mixed were allowed, newer should override older definitions
But whilst it's legitimate for there to just be old or just new, there should never be mixed entries..

I think the only change worth making, it at all, is add some to detect both and fire an error.

Reply:
I'll add a check for mixed and explicitly throw a warning and ignore legacy definitions if somehow someone has an environment with both new and legacy mixed together.